File: Assets/Resources/Lua/Base/LuaBehaviour.lua
local LuaBehaviour = class("LuaBehaviour")

function LuaBehaviour:ctor()
end

function LuaBehaviour:_createGameObject()
    local prefabPath = self:getPrefabPath()
    ResManager:CreatePrefab(
        prefabPath,
        function(go)
            self:_onGameObjectCreated(go)
        end,
        self:destroyABAfterSpawn(),
        self:destroyABAfterAllSpawnDestroy()
    )
end

function LuaBehaviour:_onGameObjectCreated(go)
    self.gameObject = go
    self.transform = go.transform
    self:onAwake()
    self:onEnable()
    local csharpLuaBehaviour = go:GetComponent(typeof(LuaFramework.LuaBehaviour))
    csharpLuaBehaviour:SetEnableAction(
        function()
            self:onEnable()
        end
    )
    csharpLuaBehaviour:SetStartAction(
        function()
            self:onStart()
        end
    )
    csharpLuaBehaviour:SetDisableAction(
        function()
            self:onDisable()
        end
    )
    csharpLuaBehaviour:SetDestroyAction(
        function()
            self:onDestroy()
        end
    )
end

--由子类重写来定义
function LuaBehaviour:getPrefabPath()
    return ""
end

--由子类重写来定义
function LuaBehaviour:destroyABAfterSpawn()
    return false
end

--由子类重写来定义
function LuaBehaviour:destroyABAfterAllSpawnDestroy()
    return false
end

function LuaBehaviour:onAwake()
    --print("-LuaBehaviour----onAwake")
end

function LuaBehaviour:onEnable()
    --print("-LuaBehaviour----onEnable")
end

function LuaBehaviour:onStart()
    --print("-LuaBehaviour----onStart")
end

function LuaBehaviour:onDisable()
    --print("-LuaBehaviour----onDisable")
end

function LuaBehaviour:onDestroy()
    --print("-LuaBehaviour----onDestroy")
end

return LuaBehaviour

File: Assets/LuaFramework/Scripts/Common/LuaBehaviour.cs
using System;
using UnityEngine;

namespace LuaFramework
{
    public class LuaBehaviour : MonoBehaviour
    {
        public string assetBundleName;
        public string prefabPath;
        public Canvas canvas { get; set; }
        public bool isFloat, destroyABAfterAllSpawnDestroy;
        Action onEnable, onStart, onDisable, onDestroy;

        /// <summary>
        /// 去掉尾部(Clone)
        /// </summary>
        string _name;

        public void SetEnableAction(Action onEnable)
        {
            this.onEnable = onEnable;
        }

        public void SetStartAction(Action onStart)
        {
            this.onStart = onStart;
        }

        public void SetDisableAction(Action onDisable)
        {
            this.onDisable = onDisable;
        }

        public void SetDestroyAction(Action onDestroy)
        {
            this.onDestroy = onDestroy;
        }

        protected virtual void Awake()
        {
            if (!canvas) canvas = GetComponent<Canvas>();
        }

        protected virtual void OnEnable()
        {
            if (onEnable != null) onEnable.Invoke();
        }

        protected virtual void Start()
        {
            if (onStart != null) onStart.Invoke();
        }

        protected virtual void OnDisable()
        {
            if (onDisable != null) onDisable.Invoke();
        }

        protected virtual void OnDestroy()
        {
            if (onDestroy != null) onDestroy.Invoke();
            UIManager.instance.OnUIDestroy(this);
            if (!string.IsNullOrEmpty(assetBundleName))
            {
                ResManager.instance.OnSpawnDestroy(assetBundleName, destroyABAfterAllSpawnDestroy);
            }
            ResManager.instance.ClearMemory();
        }
    }
}